gcp/api/server.py

```diff
@@ -110,104 +112,132 @@ def QueryAffectedBatch(self, request, context):
   @ndb_context
   def DetermineVersion(self, request, context):
     """Determine the version of the provided hashes."""
-    return determine_version(request.query, context).result()
+    res = determine_version(request.query, context).result()
+    return res
+
+
+def process_buckets(
+    file_results: List[osv.FileResult]) -> List[osv.RepoIndexBucket]:
+  """
+  Create buckets in the same process as 
+  indexer to generate the same bucket hashes
+  """
+  buckets: list[list[bytes]] = [[] for _ in range(_BUCKET_SIZE)]
+
+  for fr in file_results:
+    buckets[int.from_bytes(fr.hash[:2], byteorder='big') % _BUCKET_SIZE].append(
+        fr.hash)
+
+  results: list[osv.RepoIndexBucket] = [None] * _BUCKET_SIZE
+  for bucket_idx, bucket in enumerate(buckets):
+    bucket.sort()
+
+    hasher = hashlib.md5()
+    for v in bucket:
+      hasher.update(v)
+
+    results[bucket_idx] = osv.RepoIndexBucket(
+        node_hash=hasher.digest(),
+        files_contained=len(bucket),
+    )
+
+  return results
+
+
+def build_determine_version_result(
+    candidate_file_matches: dict[ndb.Key, int],
+    candidate_buckets: dict[ndb.Key, int], zero_match_offset: int,
+    max_files: int) -> osv_service_v1_pb2.VersionMatchList:
+  """Build sorted determine version result from the input"""
+  idx_futures = ndb.get_multi_async(candidate_file_matches.keys())
+  output = []
+
+  for f in idx_futures:
+    idx: osv.RepoIndex = f.result()
+    version_match = osv_service_v1_pb2.VersionMatch(
+        score=candidate_file_matches[idx.key] / max_files,
+        minimum_file_matches=candidate_file_matches[idx.key],
+        estimated_diff_files=estimate_diff(
```

Comment:
Not sure I fully understand the formula here just yet, but would it make sense to utilise this in the `score` calculation instead of the simple division?

Reply:
That would make a lot of sense actually, changed to have the score be based off the estimated different files value.